File: server/handler.py
import http.server
import json
import os
import tempfile
from pathlib import Path
from urllib.parse import parse_qs, urlparse

from server import projects, scan_jobs, video

STATIC_DIR = Path(__file__).resolve().parent.parent / "static"


class Handler(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/" or self.path == "":
            self._serve_file(STATIC_DIR / "index.html", "text/html")
        elif self.path.startswith("/static/"):
            rel = self.path[len("/static/") :]
            file_path = STATIC_DIR / rel
            if file_path.resolve().is_relative_to(STATIC_DIR) and file_path.is_file():
                self._serve_file(file_path)
            else:
                self._send(404, "Not found")
        elif self.path.startswith("/api/"):
            self._handle_api_get()
        else:
            self._send(404, "Not found")

    def do_POST(self):
        if self.path.startswith("/api/"):
            self._handle_api_post()
        else:
            self._send(404, "Not found")

    def do_PUT(self):
        if self.path.startswith("/api/"):
            self._handle_api_put()
        else:
            self._send(404, "Not found")

    def do_DELETE(self):
        if self.path.startswith("/api/"):
            self._handle_api_delete()
        else:
            self._send(404, "Not found")

    # --- API GET ---

    def _handle_api_get(self):
        parsed = urlparse(self.path)
        path = parsed.path
        qs = parse_qs(parsed.query)

        if path == "/api/health":
            from server.watcher import HAS_WATCHDOG

            self._json_response(
                {
                    "status": "ok",
                    "watchMode": "native" if HAS_WATCHDOG else "polling",
                    "ffmpeg": video.has_ffmpeg(),
                }
            )
        elif path == "/api/projects":
            self._json_response(projects.list_projects())
        elif path == "/api/scans":
            self._json_response(projects.list_scans())
        elif path.startswith("/api/scans/"):
            parts = path.split("/")
            if len(parts) == 5 and parts[4] == "watch":
                scan_id = parts[3]
                self._json_response({"watching": scan_jobs.is_watching(scan_id)})
            elif len(parts) == 4:
                scan_id = parts[3]
                page = int(qs.get("page", ["0"])[0])
                size = int(qs.get("size", ["50"])[0])
                status = qs.get("status", [None])[0]
                query = qs.get("q", [None])[0]
                module = qs.get("module", [None])[0]
                profile = qs.get("profile", [None])[0]
                sort = qs.get("sort", [None])[0]
                result = projects.get_scan(
                    scan_id, page, size, status, query, module, profile, sort
                )
                if result:
                    self._json_response(result)
                else:
                    self._send(404, '{"error":"scan not found"}', "application/json")
            else:
                self._send(404, '{"error":"not found"}', "application/json")
        elif path.startswith("/api/scan-jobs/"):
            parts = path.split("/")
            if len(parts) == 4:
                job = scan_jobs.get_job(parts[3])
                if job:
                    self._json_response(job)
                else:
                    self._send(404, '{"error":"job not found"}', "application/json")
            else:
                self._send(404, '{"error":"not found"}', "application/json")
        elif path == "/api/images":
            file_path = qs.get("path", [None])[0]
            if not file_path:
                self._send(400, '{"error":"path required"}', "application/json")
                return
            p = Path(file_path)
            if not p.is_file():
                self._send(404, '{"error":"file not found"}', "application/json")
                return
            if not projects.is_path_under_project(file_path):
                self._send(403, '{"error":"forbidden"}', "application/json")
                return
            self._serve_file(p)
        else:
            self._send(404, '{"error":"not found"}', "application/json")

    # --- API POST ---

    def _handle_api_post(self):
        path = urlparse(self.path).path

        if path == "/api/projects":
            body = self._read_body_json()
            if not body or "path" not in body:
                self._send(400, '{"error":"path required"}', "application/json")
                return
            name = body.get("name", Path(body["path"]).name)
            project = projects.add_project(name, body["path"])
            self._json_response(project, 201)

        elif path.startswith("/api/projects/") and path.endswith("/scan"):
            parts = path.split("/")
            if len(parts) == 5:
                project_id = parts[3]
                project = projects.get_project(project_id)
                if project:
                    job_id = scan_jobs.start_scan(
                        project_id,
                        project["name"],
                        project["path"],
                        project.get("profiles"),
                    )
                    self._json_response({"jobId": job_id}, 202)
                else:
                    self._send(404, '{"error":"project not found"}', "application/json")
            else:
                self._send(404, '{"error":"not found"}', "application/json")
        elif path.startswith("/api/scans/") and path.endswith("/watch"):
            parts = path.split("/")
            if len(parts) == 5:
                if scan_jobs.start_watching(parts[3]):
                    self._json_response({"watching": True})
                else:
                    self._send(404, '{"error":"scan not found"}', "application/json")
            else:
                self._send(404, '{"error":"not found"}', "application/json")
        elif path.startswith("/api/scan-jobs/") and path.endswith("/cancel"):
            parts = path.split("/")
            if len(parts) == 5:
                if scan_jobs.cancel_job(parts[3]):
                    self._json_response({"cancelled": True})
                else:
                    self._send(404, '{"error":"job not found"}', "application/json")
            else:
                self._send(404, '{"error":"not found"}', "application/json")
        elif path.startswith("/api/scans/") and path.endswith("/video"):
            parts = path.split("/")
            if len(parts) == 5:
                scan_id = parts[3]
                scan = projects.get_scan(scan_id, page=0, size=10000)
                if not scan:
                    self._send(404, '{"error":"scan not found"}', "application/json")
                    return
                failures = [f for f in scan.get("failures", []) if f.get("delta_path")]
                if not failures:
                    self._send(
                        400, '{"error":"no failures to export"}', "application/json"
                    )
                    return
                with tempfile.NamedTemporaryFile(suffix=".mp4", delete=False) as tmp:
                    tmp_path = tmp.name
                try:
                    video.generate_video(failures, tmp_path)
                    with open(tmp_path, "rb") as f:
                        data = f.read()
                    self.send_response(200)
                    self.send_header("Content-Type", "video/mp4")
                    self.send_header("Content-Length", str(len(data)))
                    self.send_header(
                        "Content-Disposition",
                        f'attachment; filename="papa-stud-{scan_id}.mp4"',
                    )
                    self.end_headers()
                    self.wfile.write(data)
                except Exception as e:
                    self._send(500, json.dumps({"error": str(e)}), "application/json")
                finally:
                    os.unlink(tmp_path)
            else:
                self._send(404, '{"error":"not found"}', "application/json")
        else:
            self._send(404, '{"error":"not found"}', "application/json")

    # --- API PUT ---

    def _handle_api_put(self):
        path = urlparse(self.path).path

        # PUT /api/projects/{id}/profiles
        if path.startswith("/api/projects/") and path.endswith("/profiles"):
            parts = path.split("/")
            if len(parts) == 5:
                body = self._read_body_json()
                if not body or "profiles" not in body:
                    self._send(
                        400,
                        '{"error":"profiles required"}',
                        "application/json",
                    )
                    return
                result = projects.update_project_profiles(parts[3], body["profiles"])
                if result:
                    self._json_response(result)
                else:
                    self._send(404, '{"error":"project not found"}', "application/json")
                return

        # PUT /api/scans/{id}/failures/{filename}/status
        if "/failures/" in path and path.endswith("/status"):
            parts = path.split("/")
            # /api/scans/{id}/failures/{filename}/status -> 7 parts
            if len(parts) == 7:
                scan_id = parts[3]
                filename = parts[5]
                body = self._read_body_json()
                if not body or "status" not in body:
                    self._send(400, '{"error":"status required"}', "application/json")
                    return
                stats = projects.update_failure_status(
                    scan_id, filename, body["status"]
                )
                if stats:
                    self._json_response(stats)
                else:
                    self._send(404, '{"error":"not found"}', "application/json")
                return

        # PUT /api/scans/{id}/failures/batch
        if path.endswith("/failures/batch"):
            parts = path.split("/")
            if len(parts) == 6:
                scan_id = parts[3]
                body = self._read_body_json()
                if not body or "status" not in body or "filenames" not in body:
                    self._send(
                        400,
                        '{"error":"status and filenames required"}',
                        "application/json",
                    )
                    return
                stats = projects.batch_update_status(
                    scan_id, body["filenames"], body["status"]
                )
                if stats:
                    self._json_response(stats)
                else:
                    self._send(404, '{"error":"not found"}', "application/json")
                return

        self._send(404, '{"error":"not found"}', "application/json")

    # --- API DELETE ---

    def _handle_api_delete(self):
        path = urlparse(self.path).path

        if path.startswith("/api/projects/"):
            parts = path.split("/")
            if len(parts) == 4:
                projects.delete_project(parts[3])
                self._send(204, "")
                return

        if path.startswith("/api/scans/"):
            parts = path.split("/")
            if len(parts) == 5 and parts[4] == "watch":
                scan_jobs.stop_watching(parts[3])
                self._json_response({"watching": False})
                return
            if len(parts) == 4:
                scan_jobs.stop_watching(parts[3])
                projects.delete_scan(parts[3])
                self._send(204, "")
                return

        self._send(404, '{"error":"not found"}', "application/json")

    # --- Helpers ---

    def _read_body_json(self):
        length = int(self.headers.get("Content-Length", 0))
        if length == 0:
            return None
        raw = self.rfile.read(length)
        return json.loads(raw)

    def _json_response(self, data, code=200):
        body = json.dumps(data)
        self._send(code, body, "application/json")

    def _serve_file(self, path, content_type=None):
        path = Path(path)
        if not path.is_file():
            self._send(404, "Not found")
            return
        if content_type is None:
            content_type = self._guess_type(path)
        with open(path, "rb") as f:
            data = f.read()
        self.send_response(200)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(data)))
        self.send_header("Cache-Control", "no-cache")
        self.end_headers()
        self.wfile.write(data)

    def _send(self, code, body, content_type="text/plain"):
        data = body.encode() if isinstance(body, str) else body
        self.send_response(code)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(data)))
        self.end_headers()
        self.wfile.write(data)

    @staticmethod
    def _guess_type(path):
        ext = path.suffix.lower()
        return {
            ".html": "text/html",
            ".css": "text/css",
            ".js": "application/javascript",
            ".json": "application/json",
            ".png": "image/png",
            ".jpg": "image/jpeg",
            ".jpeg": "image/jpeg",
            ".svg": "image/svg+xml",
            ".ico": "image/x-icon",
            ".woff2": "font/woff2",
        }.get(ext, "application/octet-stream")

    def log_message(self, format, *args):
        print(f"  {self.address_string()} - {format % args}")

File: src/handler.js
/**
 * Express route handlers matching the Python API surface.
 */

const express = require('express');
const path = require('path');
const fs = require('fs');
const os = require('os');
const projects = require('./projects');
const scanJobs = require('./scanJobs');
const templates = require('./templates');
const video = require('./video');
const config = require('./config');
const updateCheck = require('./updateCheck');
const apng = require('./apng');

const STATIC_DIR = path.resolve(__dirname, '..', 'static');
const INDEX_HTML = path.join(STATIC_DIR, 'index.html');

function createRouter() {
  const router = express.Router();

  // --- API GET ---

  router.get('/api/health', (req, res) => {
    res.json({ status: 'ok', watchMode: 'native', ffmpeg: video.hasFfmpeg() });
  });

  router.get('/api/update-check', async (req, res) => {
    res.json(await updateCheck.checkForUpdate());
  });

  router.get('/api/templates', (req, res) => {
    res.json(templates.listTemplates());
  });

  router.get('/api/projects', (req, res) => {
    res.json(projects.listProjects());
  });

  router.get('/api/scans', (req, res) => {
    res.json(projects.listScans());
  });

  router.get('/api/scans/:id/watch', (req, res) => {
    res.json({ watching: scanJobs.isWatching(req.params.id) });
  });

  router.get('/api/scans/:id', (req, res) => {
    const { page, size, status, q, module, profile, sort } = req.query;
    const result = projects.getScan(req.params.id, {
      page: parseInt(page) || 0,
      size: parseInt(size) || 50,
      status: status || undefined,
      query: q || undefined,
      module: module || undefined,
      profile: profile || undefined,
      sort: sort || undefined,
    });
    if (result) res.json(result);
    else res.status(404).json({ error: 'scan not found' });
  });

  router.get('/api/scan-jobs/:id', (req, res) => {
    const job = scanJobs.getJob(req.params.id);
    if (job) res.json(job);
    else res.status(404).json({ error: 'job not found' });
  });

  // Resolve a user-supplied image path against the same allowlist /api/images
  // and /api/images/meta share. Returns { real } on success or
  // { status, error } on failure.
  function resolveImagePath(filePath) {
    if (!filePath || !path.isAbsolute(filePath)) {
      return { status: 400, error: 'absolute path required' };
    }
    const ext = path.extname(filePath).toLowerCase();
    if (ext !== '.png' && ext !== '.jpg' && ext !== '.jpeg') {
      return { status: 400, error: 'only image files allowed' };
    }
    let real;
    try { real = fs.realpathSync(filePath); }
    catch { return { status: 404, error: 'not found' }; }
    const roots = [];
    for (const p of projects.listProjects()) {
      try { roots.push(fs.realpathSync(p.path)); } catch {}
    }
    try { roots.push(fs.realpathSync(path.join(projects.getDataDir(), 'cache'))); } catch {}
    if (!roots.find(r => real.startsWith(r + path.sep))) {
      return { status: 403, error: 'forbidden' };
    }
    return { real };
  }

  router.get('/api/images', (req, res) => {
    const r = resolveImagePath(req.query.path);
    if (r.error) return res.status(r.status).json({ error: r.error });
    res.sendFile(r.real);
  });

  router.get('/api/images/meta', (req, res) => {
    const r = resolveImagePath(req.query.path);
    if (r.error) return res.status(r.status).json({ error: r.error });
    res.json(apng.detectApng(r.real));
  });

  // --- Config ---

  function sendExportBundle(res, filename, bundle) {
    const safeName = filename.replace(/[^a-zA-Z0-9._-]/g, '_');
    res.set('Content-Disposition', `attachment; filename="${safeName}"`);
    res.set('Content-Type', 'application/json');
    res.send(JSON.stringify(bundle, null, 2));
  }

  router.get('/api/config/info', (req, res) => {
    const allProjects = projects.listProjects();
    const allTemplates = templates.listTemplates();
    res.json({
      dataDir: projects.getDataDir(),
      projectCount: allProjects.length,
      templateCount: allTemplates.filter(t => !t.builtin).length,
    });
  });

  router.get('/api/config/export', (req, res) => {
    const allProjects = projects.listProjects();
    const customTemplates = templates.listTemplates().filter(t => !t.builtin);
    sendExportBundle(res, 'papastud-config.json', config.createExportBundle(allProjects, customTemplates));
  });

  router.post('/api/config/import', (req, res) => {
    let bundle;
    try {
      bundle = config.migrate(req.body);
    } catch (e) {
      return res.status(400).json({ error: e.message });
    }
    let imported = { projects: 0, templates: 0 };
    if (bundle.projects && bundle.projects.length) {
      projects.importProjects(bundle.projects);
      imported.projects = bundle.projects.length;
    }
    if (bundle.templates && bundle.templates.length) {
      templates.importTemplates(bundle.templates);
      imported.templates = bundle.templates.length;
    }
    res.json(imported);
  });

  router.get('/api/config/export/project/:id', (req, res) => {
    const project = projects.getProject(req.params.id);
    if (!project) return res.status(404).json({ error: 'project not found' });
    sendExportBundle(res, `papastud-project-${project.name}.json`, config.createExportBundle([project], []));
  });

  router.get('/api/config/export/template/:id', (req, res) => {
    const t = templates.getTemplate(req.params.id);
    if (!t || t.builtin) return res.status(404).json({ error: 'custom template not found' });
    sendExportBundle(res, `papastud-template-${t.name}.json`, config.createExportBundle([], [t]));
  });

  router.post('/api/config/import/project', (req, res) => {
    const body = req.body;
    if (!body || !body.id || !body.path) return res.status(400).json({ error: 'invalid project data' });
    projects.importProjects([body]);
    res.json({ imported: 1 });
  });

  router.post('/api/config/import/template', (req, res) => {
    const body = req.body;
    if (!body || !body.id || !body.name) return res.status(400).json({ error: 'invalid template data' });
    templates.importTemplates([body]);
    res.json({ imported: 1 });
  });

  router.post('/api/config/reset', (req, res) => {
    projects.resetProjects();
    templates.resetTemplates();
    res.json({ reset: true });
  });

  // --- API POST ---

  router.post('/api/projects', (req, res) => {
    const body = req.body;
    if (!body || !body.path) return res.status(400).json({ error: 'path required' });
    const expanded = projects.expandHome(body.path);
    if (!path.isAbsolute(expanded)) {
      return res.status(400).json({ error: `absolute path required: ${body.path}` });
    }
    const resolved = path.resolve(expanded);
    const shown = resolved !== body.path ? `${body.path} (expanded to ${resolved})` : body.path;
    let stat;
    // codeql[js/path-injection]: by design — this single-user app lets the user
    // pick any directory on their machine to scan. Server binds to 127.0.0.1
    // (Electron only; no remote attacker). The path is normalized via
    // expandHome + path.resolve and rejected if not absolute.
    try { stat = fs.statSync(resolved); }
    catch { return res.status(400).json({ error: `path does not exist: ${shown}` }); }
    if (!stat.isDirectory()) return res.status(400).json({ error: `path is not a directory: ${shown}` });
    const name = body.name || path.basename(resolved);
    const project = projects.addProject(name, resolved, body.template_ids, body.strategy, {
      xcresult_path: body.xcresult_path,
    });
    res.status(201).json(project);
  });

  router.post('/api/templates', (req, res) => {
    const body = req.body;
    if (!body || !body.name) return res.status(400).json({ error: 'name required' });
    const t = templates.createTemplate(body);
    res.status(201).json(t);
  });

  router.post('/api/projects/:id/scan', (req, res) => {
    const project = projects.getProject(req.params.id);
    if (!project) return res.status(404).json({ error: 'project not found' });
    const jobId = scanJobs.startScan(project);
    res.status(202).json({ jobId });
  });

  router.post('/api/scans/:id/watch', (req, res) => {
    if (scanJobs.startWatching(req.params.id)) {
      res.json({ watching: true });
    } else {
      res.status(404).json({ error: 'scan not found' });
    }
  });

  router.post('/api/scan-jobs/:id/cancel', (req, res) => {
    if (scanJobs.cancelJob(req.params.id)) {
      res.json({ cancelled: true });
    } else {
      res.status(404).json({ error: 'job not found' });
    }
  });

  router.post('/api/scans/:id/video', async (req, res) => {
    const scanId = req.params.id.replace(/[^a-zA-Z0-9._-]/g, '');
    const { module: mod, profile, q } = req.query;
    const scan = projects.getScan(scanId, {
      page: 0, size: 10000,
      module: mod || undefined,
      profile: profile || undefined,
      query: q || undefined,
    });
    if (!scan) return res.status(404).json({ error: 'scan not found' });
    const failures = (scan.failures || []).filter(f => f.delta_path);
    if (!failures.length) return res.status(400).json({ error: 'no failures to export' });

    const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'papastud-vid-'));
    const tmpFile = path.join(tmpDir, 'output.mp4');
    try {
      await video.generateVideo(failures, tmpFile);
      const filterTag = [mod, profile, q].filter(Boolean).join('-').replace(/[^a-zA-Z0-9._-]/g, '') || 'all';
      res.set('Content-Disposition', `attachment; filename="papa-stud-${scanId}-${filterTag}.mp4"`);
      res.set('Content-Type', 'video/mp4');
      const stat = fs.statSync(tmpFile);
      res.set('Content-Length', String(stat.size));
      const stream = fs.createReadStream(tmpFile);
      stream.on('error', () => {
        fs.rmSync(tmpDir, { recursive: true, force: true });
        if (!res.headersSent) res.status(500).end();
      });
      stream.pipe(res).on('finish', () => {
        fs.rmSync(tmpDir, { recursive: true, force: true });
      });
    } catch (e) {
      console.error('Video generation failed:', e.message);
      fs.rmSync(tmpDir, { recursive: true, force: true });
      if (!res.headersSent) res.status(500).json({ error: e.message });
    }
  });

  // --- API PUT ---

  router.put('/api/projects/:id/profiles', (req, res) => {
    const body = req.body;
    if (!body || !body.profiles) return res.status(400).json({ error: 'profiles required' });
    const result = projects.updateProjectProfiles(req.params.id, body.profiles);
    if (result) res.json(result);
    else res.status(404).json({ error: 'project not found' });
  });

  router.put('/api/scans/:scanId/failures/:filename/status', (req, res) => {
    const body = req.body;
    if (!body || !body.status) return res.status(400).json({ error: 'status required' });
    const stats = projects.updateFailureStatus(req.params.scanId, req.params.filename, body.status);
    if (stats) res.json(stats);
    else res.status(404).json({ error: 'not found' });
  });

  router.put('/api/scans/:scanId/failures/batch', (req, res) => {
    const body = req.body;
    if (!body || !body.status || !body.filenames) {
      return res.status(400).json({ error: 'status and filenames required' });
    }
    const stats = projects.batchUpdateStatus(req.params.scanId, body.filenames, body.status);
    if (stats) res.json(stats);
    else res.status(404).json({ error: 'not found' });
  });

  router.post('/api/scans/:scanId/failures/:filename/accept', (req, res) => {
    const result = projects.acceptBaseline(req.params.scanId, req.params.filename);
    if (!result) return res.status(404).json({ error: 'not found' });
    if (result.error) return res.status(400).json({ error: result.error });
    res.json(result);
  });

  router.post('/api/scans/:scanId/accept-all', (req, res) => {
    const result = projects.acceptAllBaselines(req.params.scanId);
    if (!result) return res.status(404).json({ error: 'scan not found' });
    if (result.error) return res.status(400).json({ error: result.error });
    res.json(result);
  });

  // --- API DELETE ---

  router.delete('/api/templates/:id', (req, res) => {
    if (templates.deleteTemplate(req.params.id)) {
      res.status(204).end();
    } else {
      res.status(400).json({ error: 'cannot delete built-in template' });
    }
  });

  router.delete('/api/projects/:id', (req, res) => {
    projects.deleteProject(req.params.id);
    res.status(204).end();
  });

  router.delete('/api/scans/:id/watch', (req, res) => {
    scanJobs.stopWatching(req.params.id);
    res.json({ watching: false });
  });

  router.delete('/api/scans/:id', (req, res) => {
    scanJobs.stopWatching(req.params.id);
    projects.deleteScan(req.params.id);
    res.status(204).end();
  });

  return router;
}

function createApp() {
  const app = express();
  app.use(express.json());
  app.use(createRouter());
  app.use('/static', express.static(STATIC_DIR));
  // Serve index.html from a known constant path
  app.get('/', (req, res) => res.sendFile(INDEX_HTML));
  return app;
}

module.exports = { createRouter, createApp, STATIC_DIR };
